File: compiler-core/src/main/java/com/siimkinks/sqlitemagic/validator/TransformerValidator.java
package com.siimkinks.sqlitemagic.validator;

import com.google.common.base.Joiner;
import com.siimkinks.sqlitemagic.Environment;
import com.siimkinks.sqlitemagic.annotation.Table;
import com.siimkinks.sqlitemagic.annotation.transformer.DbValueToObject;
import com.siimkinks.sqlitemagic.annotation.transformer.ObjectToDbValue;
import com.siimkinks.sqlitemagic.annotation.transformer.Transformer;
import com.siimkinks.sqlitemagic.element.ExtendedTypeElement;
import com.siimkinks.sqlitemagic.element.TransformerElement;

import javax.inject.Inject;
import javax.inject.Singleton;
import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;

import static com.siimkinks.sqlitemagic.Const.ERR_MULTIPLE_TRANSFORMERS;
import static com.siimkinks.sqlitemagic.Const.ERR_TRANSFORM_TYPES_NOT_MATCHING;
import static com.siimkinks.sqlitemagic.Const.SQL_TYPE_MAP;
import static com.siimkinks.sqlitemagic.Const.SUPPORTED_SQL_BOXED_TYPES;

@Singleton
public class TransformerValidator {

  final Environment environment;

  @Inject
  public TransformerValidator(Environment environment) {
    this.environment = environment;
  }

  public boolean isTransformerValid(TransformerElement transformerElement) {
    final Element blameElement = transformerElement.getRandomBlameElement();
    if (transformerElement.isMissingMethods()) {
      environment.error(blameElement, "There must be 2 static annotated transform methods -- " +
              "@%s annotated method which transforms Java objects to SQLite compatible objects and " +
              "@%s annotated method which transforms database values to Java objects.\n" +
              "%s",
          ObjectToDbValue.class.getSimpleName(),
          DbValueToObject.class.getSimpleName(),
          transformerElement.missingMethodsErrorInfo());
      return false;
    }
    if (!isTransformerMethodValid(environment, transformerElement.getObjectToDbValueMethod(), transformerElement.getDbValueToObjectMethod())) {
      return false;
    }
    if (!isTransformerMethodValid(environment, transformerElement.getDbValueToObjectMethod(), transformerElement.getObjectToDbValueMethod())) {
      return false;
    }
    final TypeElement serializedType = transformerElement.getSerializedType().getTypeElement();
    if (serializedType == null) {
      environment.error(blameElement, "Unsupported serialization type %s", transformerElement.getRawSerializedType().toString());
      return false;
    }
    final ExtendedTypeElement deserializedType = transformerElement.getDeserializedType();
    if (deserializedType.getTypeElement() == null) {
      environment.error(blameElement, "Unsupported deserialization type %s", transformerElement.getRawDeserializedType().toString());
      return false;
    }
    final String deserializedQualifiedName = deserializedType.getQualifiedName();
    if (SQL_TYPE_MAP.get(deserializedQualifiedName) != null) {
      environment.error(blameElement, "SQL types [%s] can't have transformers",
          Joiner.on(", ").join(SUPPORTED_SQL_BOXED_TYPES));
      return false;
    }
    final String serializedTypeQualifiedName = Environment.getQualifiedName(serializedType);
    if (SQL_TYPE_MAP.get(serializedTypeQualifiedName) == null) {
      environment.error(blameElement,
          "Provided serialization type was %s, but serialized type must be one of supported SQLite types - %s",
          serializedTypeQualifiedName,
          Joiner.on(", ").join(SUPPORTED_SQL_BOXED_TYPES));
      return false;
    }
    if (environment.hasTransformerFor(deserializedType)) {
      environment.error(blameElement, ERR_MULTIPLE_TRANSFORMERS + deserializedQualifiedName);
      return false;
    }
    if (deserializedType.getTypeElement().getAnnotation(Table.class) != null) {
      environment.error(blameElement, "Cannot transform object %s which is also annotated with @%s. Delete transformer or remove annotation",
          deserializedQualifiedName, Table.class.getSimpleName());
      return false;
    }
    return true;
  }

  private static boolean isTransformerMethodValid(Environment environment,
                                                 ExecutableElement method,
                                                 ExecutableElement otherMethod) {
    if (method.getParameters().size() != 1) {
      environment.error(method, "Transformer methods must have one parameter");
      return false;
    }
    if (!method.getModifiers().contains(Modifier.STATIC)) {
      environment.error(method, "Transformer methods must be static");
      return false;
    }
    if (otherMethod != null) {
      if (!assertReturnTypeSameAsFirstParamType(environment, method, otherMethod)) {
        return false;
      }
      if (!assertReturnTypeSameAsFirstParamType(environment, otherMethod, method)) {
        return false;
      }
    }
    return true;
  }

  private static boolean assertReturnTypeSameAsFirstParamType(Environment environment, ExecutableElement firstMethod, ExecutableElement secondMethod) {
    TypeMirror firstParam = firstMethod.getParameters().get(0).asType();
    TypeMirror returnType = secondMethod.getReturnType();
    if (!environment.getTypeUtils().isSameType(firstParam, returnType)) {
      environment.error(firstMethod, ERR_TRANSFORM_TYPES_NOT_MATCHING);
      return false;
    }
    return true;
  }
}

File: compiler-ksp/src/main/kotlin/com/siimkinks/sqlitemagic/transformer/TransformerValidator.kt
package com.siimkinks.sqlitemagic.transformer

import com.google.devtools.ksp.isAnnotationPresent
import com.google.devtools.ksp.symbol.ClassKind.OBJECT
import com.google.devtools.ksp.symbol.FunctionKind.STATIC
import com.google.devtools.ksp.symbol.FunctionKind.TOP_LEVEL
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.Modifier.PRIVATE
import com.google.devtools.ksp.symbol.Modifier.SUSPEND
import com.siimkinks.sqlitemagic.Const.SUPPORTED_TRANSFORMER_DESERIALIZED_TYPES
import com.siimkinks.sqlitemagic.Environment
import com.siimkinks.sqlitemagic.annotation.Table
import com.siimkinks.sqlitemagic.annotation.transformer.DbValueToObject
import com.siimkinks.sqlitemagic.annotation.transformer.ObjectToDbValue

class TransformerValidator(
  private val environment: Environment
) {
  fun isTransformerValid(transformer: TransformerRoundElement): Boolean {
    val blameSymbol = transformer.randomBlameSymbol
    if (transformer.isMissingMethods) {
      environment.logger.error(
        message = buildString {
          append("There must be 2 annotated transform methods callable without an instance -- ")
          append("@${ObjectToDbValue::class.simpleName} annotated method with single parameter which transforms objects to SQLite compatible objects and ")
          append("@${DbValueToObject::class.simpleName} annotated method with single parameter which transforms database values to objects.\n")
          append(transformer.missingMethodsErrorInfo())
        },
        symbol = blameSymbol
      )
      return false
    }

    val objectToDbValueMethod = checkNotNull(transformer.objectToDbValueMethod)
    val dbValueToObjectMethod = checkNotNull(transformer.dbValueToObjectMethod)
    if (objectToDbValueMethod.declaration == dbValueToObjectMethod.declaration) {
      environment.logger.error(
        message = "There must be 2 annotated transform methods; one method cannot serve both transformation directions",
        symbol = blameSymbol
      )
      return false
    }
    if (!isTransformerMethodValid(method = objectToDbValueMethod)) {
      return false
    }
    if (!isTransformerMethodValid(method = dbValueToObjectMethod)) {
      return false
    }
    if (!assertReturnTypeSameAsFirstParamType(firstMethod = objectToDbValueMethod, secondMethod = dbValueToObjectMethod)) {
      return false
    }
    if (!assertReturnTypeSameAsFirstParamType(firstMethod = dbValueToObjectMethod, secondMethod = objectToDbValueMethod)) {
      return false
    }

    val serializedType = transformer.serializedType
    if (serializedType?.declaration == null) {
      environment.logger.error(
        message = "Unsupported serialization type ${transformer.serializedType?.typeKey}",
        symbol = blameSymbol
      )
      return false
    }
    val deserializedType = transformer.deserializedType
    if (deserializedType?.declaration == null) {
      environment.logger.error(
        message = "Unsupported deserialization type ${transformer.deserializedType?.typeKey}",
        symbol = blameSymbol
      )
      return false
    }
    if (deserializedType.sqlStorageType != null) {
      environment.logger.error(
        message = "SQL types [${SUPPORTED_TRANSFORMER_DESERIALIZED_TYPES.joinToString()}] can't have transformers",
        symbol = blameSymbol
      )
      return false
    }
    if (serializedType.sqlStorageType == null) {
      environment.logger.error(
        message = "Provided serialization type was ${serializedType.typeName}, but serialized type must be one of supported SQLite types - " +
            SUPPORTED_TRANSFORMER_DESERIALIZED_TYPES.joinToString(),
        symbol = blameSymbol
      )
      return false
    }
    val existingTransformer = environment.getTransformerFor(deserializedType)
    if (existingTransformer != null && existingTransformer != transformer.toTransformerElement()) {
      environment.logger.error(
        message = "Multiple transformers defined for ${deserializedType.qualifiedName}",
        symbol = blameSymbol
      )
      return false
    }
    if (deserializedType.declaration.isAnnotationPresent(Table::class)) {
      environment.logger.error(
        message = "Cannot transform object ${deserializedType.qualifiedName} which is also annotated with @Table. Delete transformer or remove annotation",
        symbol = blameSymbol
      )
      return false
    }
    return true
  }

  private fun isTransformerMethodValid(method: TransformerRoundMethodElement): Boolean {
    if (!method.hasSingleParameter) {
      environment.logger.error(
        message = "Transformer methods must have one parameter",
        symbol = method.declaration
      )
      return false
    }
    if (method.declaration.extensionReceiver != null) {
      environment.logger.error(
        message = "Transformer methods must not be extension functions",
        symbol = method.declaration
      )
      return false
    }
    if (SUSPEND in method.declaration.modifiers) {
      environment.logger.error(
        message = "Transformer methods must not be suspend functions",
        symbol = method.declaration
      )
      return false
    }
    if (PRIVATE in method.declaration.modifiers) {
      environment.logger.error(
        message = "Transformer methods must not be private",
        symbol = method.declaration
      )
      return false
    }
    if (!method.declaration.isTransformerCallableWithoutInstance()) {
      environment.logger.error(
        message = "Transformer methods must be top-level, declared in an object or companion object, or Java static",
        symbol = method.declaration
      )
      return false
    }
    return true
  }

  private fun assertReturnTypeSameAsFirstParamType(
    firstMethod: TransformerRoundMethodElement,
    secondMethod: TransformerRoundMethodElement
  ): Boolean {
    val firstParamType = firstMethod.firstParameterType
    val returnType = secondMethod.returnType
    if (returnType == null || firstParamType?.type != returnType.type) {
      environment.logger.error(
        message = "One transformer method's return type must be the same as other method's first parameter and vice versa",
        symbol = firstMethod.declaration
      )
      return false
    }
    return true
  }
}

private fun KSFunctionDeclaration.isTransformerCallableWithoutInstance(): Boolean {
  if (functionKind == TOP_LEVEL || functionKind == STATIC) {
    return true
  }
  val parentClass = parentDeclaration as? KSClassDeclaration ?: return false
  return parentClass.classKind == OBJECT || parentClass.isCompanionObject
}
